### PR TITLE
Make deleted code in a suggestion clearer

### PR DESCRIPTION
**Pull request by estebank**
Show suggestions that include deletions in a way similar to `diff`'s output.

For changes that do not have deletions, use `+` as an underline for additions and `~` as an underline for replacements.

For multiline suggestions, we now use `~` in the gutter to signal replacements and `+` to signal whole line replacements/additions.

In all cases we now use color to highlight the specific spans and snippets.

### CODE REVIEW

File: src/test/ui/suggestions/match-prev-arm-needing-semi.stderr [hunk 1]
```diff
@@ -28,7 +28,7 @@ LL |         false => async_dummy().await,
 help: consider removing this semicolon
    |
 LL |             async_dummy()
-   |                         --
+   |                         /\
```
Comment (JohnTitor) [inline src/test/ui/suggestions/match-prev-arm-needing-semi.stderr:31]: This makes me think that underline + extra space might be better than `/\`...

File: src/test/ui/single-use-lifetime/one-use-in-trait-method-argument.stderr [hunk 2]
```diff
@@ -14,7 +14,7 @@ LL | #![deny(single_use_lifetimes)]
 help: elide the single-use lifetime
    |
 LL |     fn next(&mut self) -> Option<Self::Item> {
-   |           ----
+   |           /\/\
```
Comment (JohnTitor) [inline src/test/ui/single-use-lifetime/one-use-in-trait-method-argument.stderr:17]: This case seems better than the current. The current is hard to notice that we remove two `'g`s, yeah.

File: src/test/ui/traits/bound/not-on-struct.stderr [hunk 3]
```diff
@@ -118,7 +118,7 @@ LL | fn e() -> 'static + A + 'static {
 help: if you meant to use a type and not a trait here, remove the bounds
    |
 LL | fn e() -> A {
-   |          ---
+   |          //\
```
Comment (JohnTitor) [inline src/test/ui/traits/bound/not-on-struct.stderr:121]: Looks like there's an odd case if one span overlaps another. Uhm, I'd prefer extra spaces here as well.

**Comment by Nicholas-Baron** [comment]
I personally view the narrow end of an arrow as a destination to insert some items.
Thus, the bottom line of that error message would something look like
```
fn foo(&self, _ : &impl Debug) { }
     \/
```

However, I do get the choice of `/\` from an aesthetic perspective (e.g. looks like the `^`).
Another idea may be using red for the `/\` to signify deletion, which seems to be a common cross-platform idiom.

**Comment by iago-lito** [comment]
Just thinking.. what about making it clearer with underline+overline, to avoid @Nicholas-Baron "insert location" interpretation?
```
      \/
fn next(&mut self)
      /\
```

**Comment by jieyouxu** [comment]
Adding to the bikeshed...

```
help: try removing the generic parameter and using `impl Trait` instead
  |
8 |         fn foo(&self, _: &impl Debug) { }
  |              --           ++++++++++
```

Red for deletions `-`, green for additions `+`, perhaps like diffs? This has the advantage of showing clearly what's added and should be easier to compute the width for, though the `-` symbol is already to highlight.

Screen Shot 2021-06-22 at 20 15 19

Maybe use `x` to signal deletions?

Screen Shot 2021-06-22 at 20 11 56

Maybe even highlight what the user should try to add?

Screen Shot 2021-06-22 at 20 12 34

**Comment by bjorn3** [comment]
> Maybe even highlight what the user should try to add?

The code used to be the same color as the highlights below it, but was changed to be the default color at some point.

**Comment by camsteffen** [comment]
Another idea:
image

**Comment by bjorn3** [comment]
That doesn't work without colors. Think about a color blind person or copy pasted error messages.

**Comment by camsteffen** [comment]
I think it works without colors.

```text
 help: try removing the generic parameter and using `impl Trait` instead
   |
 8 |         fn foo(&self, _: &impl Debug){ }
   |              ><          >++++++++++<

**Comment by bjorn3** [comment]
Oh, the `>` and `<` are outside of the inserted range instead of inside. That is somewhat non-obvious.

**Comment by estebank** [comment]
The other option is to keep the removed code and give it a red background or color, and a different underline, like was proposed above so that it still works with colorless output:

```
help: try removing the generic parameter and using `impl Trait` instead
  |
8 |         fn foo<U: Debug>(&self, _: &impl Debug) { }
  |               xxxxxxxxxx            ++++++++++
```

**Comment by camsteffen** [comment]
> Oh, the `>` and `<` are outside of the inserted range instead of inside. That is somewhat non-obvious.

Yeah. On second thought I don't think I would use `>+<` for inserted code.

I think that `><` may be more intuitive than `/\` for deleted code. Neither seems clearly better to me.

> keep the removed code

I don't like that option personally. I would like the output to show exactly what the code looks like after applying the suggestion. I can compare with the current version of the code in the first error message.

**Comment by tlyu** [comment]
I think `/\` looks too much like a [proofreader's mark for insertion](https://www.chicagomanualofstyle.org/help-tools/proofreading-marks.html), so using it to signify deletion seems like a bad idea. Also, I'm not sure it always renders as a legible arrow or caret type appearance in all likely terminal fonts.

I kind of like `><` because it looks like it's pointing to the empty space where the deleted text used to be.

There seems to be a [suggestion for electronic texts](https://people.wku.edu/sally.kuhlenschmidt/proofread.htm) to use marks including `%` to signify deletion, because of the visual similarity to the traditional handwritten proofreader's mark. For our case, that would have to go inline with the old text that we're suggesting deleting, which is probably not something we want.

Another possibility is to use inline comments like `/* deleted */` to signify the deletion, or even just surround the deleted text with an inline comment. Drawbacks include less familiarity with this style of comment (it doesn't seem to be the prevailing style, and rarely appears in tutorial material) and repetition of the deleted text (if we use the "comment out the deleted text" approach).

**Comment by estebank** [comment]
> I would like the output to show exactly what the code looks like after applying the suggestion. I can compare with the current version of the code in the first error message.

That's been my personal take, but I've encountered enough people that have gotten confused by it that it's making me reconsider my stance for the projects sake.

---

The problem I see with using `+` for the underline is that if we are _replacing_ text it doesn't convey the right message, as we are not informing the user that the replacement occurred. That being said:

Screen Shot 2021-06-22 at 11 04 02 AM

**Comment by camsteffen** [comment]
> The problem I see with using `+` for the underline is that if we are _replacing_ text it doesn't convey the right message, as we are not informing the user that the replacement occurred.

I'm convinced on that. It looks like diff output but the semantics are different. That's confusing.

**Comment by Artoria2e5** [comment]
While we’re at it with the red mark, maybe a strikethrough can be tried too? The \\e[9 stuff, not the Unicode hack.

**Comment by estebank** [comment]
> While we’re at it with the red mark, maybe a strikethrough can be tried too? The \e[9 stuff, not the Unicode hack.

We use termcolor which doesn't have support to specify strikethrough.

Using unicode, I have this mockup. Please ignore the obvious bugs with color positions:

Screen Shot 2021-06-22 at 11 56 08 AM

I don't know if the strikethrough will help that much.

**Comment by tlyu** [comment]
> While we’re at it with the red mark, maybe a strikethrough can be tried too? The \e[9 stuff, not the Unicode hack.

I think that ANSI terminal strikethrough capability isn't widely supported? Also, I think it's important that any annotations survive copying and pasting, especially because we often ask people to copy and paste full error messages when they ask about compiler errors.

**Comment by tlyu** [comment]
> > The problem I see with using `+` for the underline is that if we are _replacing_ text it doesn't convey the right message, as we are not informing the user that the replacement occurred.
> 
> I'm convinced on that. It looks like diff output but the semantics are different. That's confusing.

I agree. Maybe `~` as an underline for changes? (It's similar to the proofreader's mark for transposition, which is a bit of a stretch but close enough, I think.)

**Comment by tlyu** [comment]
>> I would like the output to show exactly what the code looks like after applying the suggestion. I can compare with the current version of the code in the first error message.
>
> That's been my personal take, but I've encountered enough people that have gotten confused by it that it's making me reconsider my stance for the projects sake.

I think if we're going to leave the deleted text inline, I'm leaning toward surrounding it with a `/* */` comment.

**Comment by estebank** [comment]
Three other options. I'm partial to the second one, but the first one _looks_ easier to understand what's going on.

Screen Shot 2021-06-22 at 4 10 49 PM
Screen Shot 2021-06-22 at 4 17 49 PM
Screen Shot 2021-06-22 at 4 24 11 PM

**Comment by estebank** [comment]
Current output for this PR:

Screen Shot 2021-06-22 at 4 58 54 PM

**Comment by JohnTitor** [comment]
> Three other options. I'm partial to the second one, but the first one looks easier to understand what's going on.

I'm worried that how the first looks like with `--color=never`. The second and the current look nice to me.

**Comment by jieyouxu** [comment]
> Current output for this PR:
> 
> Screen Shot 2021-06-22 at 4 58 54 PM

I also like this current output. I would like to summarize some UX considerations presented in the
discussions prior that I believe contributes to good suggestion UX:

- The replacement suggestion should be **valid Rust code** (or at least a step towards that
  direction) which guides the user towards a proper solution. This is the case with not leaving the
  removed spans behind in the suggestion.
- The deletion sigil used (`~`) corresponds well to spell check errors (squiggly/wave-y underlines)
  commonly used by word processors, editors and browsers to hint that a "suggested replacement" is
  available; this can perhaps be more newcomer-friendly due to **familiarity**. I also think that
  `-` might collide with its use to indicate location; can be confusing especially without colors. I
  also agree that `+` can be ambiguous *cf.* diffs.

<p align="center">
	
</p>

- Different sigils are used (`~` for deletion, `^` for addition) that are able to clearly
  distinguish between additions and deletions **without relying on colors** (friendly for those who
  are colorblind). For example:

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
8 |         fn foo  (&self, _: &impl Debug) { }
  |               ~~            ^^^^^^^^^^
```

- **Simple ASCII characters** used without exotic whitespace or other symbols; easy for user to type
  out and copy, wide range of terminal/editor/IDE support. The additional advantage of having
  simpler underlines such as `~~~` and `^^^` is that it should be easier to compute their respective
  spans.
- The selected sigils also **works next to each other** reasonably well (*contrived example*...
  since currently presented mockups don't have addition/deletion spans next to each other):

```
help:  an interesting error message
  |
2 |     let x: u64 = 0;
  |             ~^
```

(On a side note, how would the addition/deletion underlines be handled if their spans *overlap* or
are *neighbouring*?)

**Comment by bjorn3** [comment]
> ```
> help:  an interesting error message
>   |
> 2 |     let x: u64 = 0;
>   |             ~^
> ```

Gcc uses the exact same squiggles for a different purpose:

```
<source>: In function 'int square(int)':
<source>:3:18: error: 'foo' was not declared in this scope
    3 |     return num * foo;
      |                  ^~~
```

or

```
<source>: In function 'int square(int)':
<source>:5:16: error: no match for 'operator*' (operand types are 'int' and 'Foo')
    5 |     return num * Foo{};
      |            ~~~ ^ ~~~~~
      |            |     |
      |            int   Foo
```

For reference this is what gcc uses for insertions:

```
<source>:5:16: error: stray '@' in program
    5 |     return num @ Foo{};
      |                ^
<source>: In function 'int square(int)':
<source>:5:15: error: expected ';' before 'Foo'
    5 |     return num @ Foo{};
      |               ^  ~~~
      |               ;
```

and clang:

```
<source>:5:15: error: expected ';' after return statement
    return num @ Foo{};
              ^
              ;
1 error generated.
```

**Comment by iago-lito** [comment]
I agree with the following principles to stick to:
- __A.__ Replacement suggestion is valid rust code 
  - __A.1__ If possible well-formatted.
- __B__ Deletion and insertion can be read without colors, colors is only a bonus.
- __C__ No fancy unicode chars or terminal strikethrough.
- __D__ Attach neat semantics to the chars used to delimitate deleted/added/changed regions
  - __D.1__ Yet don't conflict with usual diff semantics.

Then what about :
regular

With red `><` for deletion, blue `~` for edition and green `^` for addition.

The only problem I see is possible overlap, in which case I would only derogate to __A.1__ and fix the overlap with additional whitespace, `e.g.`:
Additional whitespace

**Comment by estebank** [comment]
This looks a bit crowded, but could work
Screen Shot 2021-06-23 at 12 49 33 PM

**Comment by jieyouxu** [comment]
> This looks a bit crowded, but could work
> Screen Shot 2021-06-23 at 12 49 33 PM

While this version aligns well with diff lines `-`/`+`, I think it relies on color otherwise its utility trade-off for another line seems limited?

I.e. without colors, that mockup becomes

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
  -         fn foo<U: Debug>(&self, _: &U) { }
8 +         fn foo(&self, _: &impl Debug) { }
  |              ><           ^^^^^^^^^^
```

(Also should they be annotated with the same line number to avoid confusion?)

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
8 -         fn foo<U: Debug>(&self, _: &U) { }
8 +         fn foo(&self, _: &impl Debug) { }
  |              ><           ^^^^^^^^^^
```

I think another thing that might have been overlooked in the previous mockups is e.g. in the `&U` => `&impl Debug`, only the insertion of `impl Debug` is signified by an insertion sigil but not the deletion of the `U` type parameter with an deletion sigil, i.e.

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
8 |         fn foo  (&self, _: & impl Debug) { }
  |               ~~            ~^^^^^^^^^^
```

here `<U: Debug>` is removed, and thus is signified by deletion sigils, but `U` in `&U` is also removed but is not signified likewise. This omission/asymmetry is clearly revealed by the proposed diff-lines approach.

Going with the quoted output, but with insertion/deletion sigils fully respecting what's added and removed:

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
  -         fn foo<U: Debug>(&self, _: &U) { }
8 +         fn foo  (&self, _: &  impl Debug) { }
  |               ><            ><^^^^^^^^^^
```

Note there that if `><` is used to signify deletion, it is two characters wide and it may be clearer if a space placeholder is used instead of trying to align it to the left - see the `><` at the `&impl Debug` position because a single `U` is removed but `><` is two characters wide and `><` neighbours `^^^`?

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
  -         fn foo<U: Debug>(&self, _: &U) { }
8 +         fn foo(&self, _: &impl Debug) { }
  |              ><          ><^^^^^^^^^^
```

**Comment by camsteffen** [comment]
Using `~` for deletion seems odd. I don't think it is any better than `-`. But using `~` for *replacement* makes more sense to me. Perhaps that is more analogous to a spell checker.

Adding space(s) in place of deleted code detracts from the intuitiveness of `><` so I would say these two things (added spaces and `><`) are mutually exclusive. I think `><` with no spaces would be my first choice, but I see the benefit of adding spaces too (especially with the red highlighting).

**Comment by tlyu** [comment]
I agree with using `~` for changes and `^` for insertions. `><` for deletions also seems reasonable.

I would also think that if we're going to show before and after versions, maybe adding an additional line for markings ("overlines"?) would be helpful? Then we could use another sigil like `%` above deleted characters, and also indicate the spans of the "before" version of changed text:

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
  |               %%%%%%%%%%           ~~
8 -         fn foo<U: Debug>(&self, _: &U) { }
8 +         fn foo(&self, _: &impl Debug) { }
  |              ><          ~~~~~~~~~~~
```

I think merging adjacent or overlapping deletions and additions (or changes) into a single "changed" span might be best, instead of adding additional spacing.

There are probably disadvantages to adding yet occupying another line of text with the overlines, though.

**Comment by estebank** [comment]
I tried using an overline (it's in one of the commits here) but it doesn't convince me. It's not terrible, but some of the feedback has been quite negative about it.

Screen Shot 2021-06-24 at 3 33 27 PM

I'm also looking at what it would be like without an _underline_
Screen Shot 2021-06-24 at 4 39 42 PM

File: src/test/ui/suggestions/match-prev-arm-needing-semi.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/suggestions/match-prev-arm-needing-semi.stderr]: Changed the output radically, please take a look at the new `diff`-like format.

**Comment by jieyouxu** [comment]
> I tried using an overline (it's in one of the commits here) but it doesn't convince me. It's not terrible, but some of the feedback has been quite negative about it.
> 
> Screen Shot 2021-06-24 at 3 33 27 PM

The overline seems somehow *more* confusing to me, especially since it doesn't seem to be consistent with underlines used in other error/diagnostic messages... And it adds even more lines?

> I'm also looking at what it would be like without an _underline_
> Screen Shot 2021-06-24 at 4 39 42 PM

This seems quite clear IMO, with one caveat - the user seems to need to rely on the color to derive the modified spans, i.e. without color:

```
help:  try removing the generic parameter and using `impl Trait` instead
  |
8 -         fn foo<U: Debug>(&self, _: &U) { }
8 +         fn foo(&self, _: &impl Debug) { }
  |
```

without underlines or some sort of non-color marker, the user doesn't have a good indication of modified/added/deleted spans? E.g. imagine copy-pasting as plain-text, or for those who are colorblind, especially those who are unable to clearly distinguish between the red/green colors.

**Comment by iago-lito** [comment]
I also find the overline confusing, and wish that color be not the only cue.. but I have to admit that there is a tension between `><` marker and the possibility of overlap.

Maybe overline can be used as a last resort to fix overlapping cases, which are maybe rare anyway?

overline to fix overlap

Or we need to fix overlap with additional whitespace

whitespace to fix overlap

Making `><` "eat" a character from the `~~~~~~~` is not a solution either, because there would be ambiguous cases (think of `function<T>(&self)` being changed to `f(&self)`, then there would be no room for the `~` anymore).

In the end, as long as we cannot *slip* a sigil in-between two characters to represent deletion (we can't, right?), I doubt there are other neat solutions to handle overlap. I think that the whitespace approach is the less confusing, although it does break requirement [__A.1__](https://github.com/rust-lang/rust/pull/86532#issuecomment-866643909).

**Comment by jieyouxu** [comment]
> I also find the overline confusing, and wish that color be not the only cue.. but I have to admit that there is a tension between `><` marker and the possibility of overlap.
> 
> Maybe overline can be used as a last resort to fix overlapping cases, which are maybe rare anyway?
> 
> overline to fix overlap
> 
> Or we need to fix overlap with additional whitespace
> 
> whitespace to fix overlap
> 
> Making `><` "eat" a character from the `~~~~~~~` is not a solution either, because there would be ambiguous cases (think of `function<T>(&self)` being changed to `f(&self)`, then there would be no room for the `~` anymore).
> 
> In the end, as long as we cannot *slip* a sigil in-between two characters to represent deletion (we can't, right?), I doubt there are other neat solutions to handle overlap. I think that the whitespace approach is the less confusing, although it does break requirement [__A.1__](https://github.com/rust-lang/rust/pull/86532#issuecomment-866643909).

I think being well formatted isn't as important as having hopefully valid Rust code while not being ambiguous, but that's just my opinion. It'd be great if it doesn't need an additional rustfmt run though.

**Comment by estebank** [comment]
> the user seems to need to rely on the color to derive the modified spans

It doesn't _rely_ on the color to identify the differences, the color _makes it easier_, but you can still "compare and contrast" the two lines. The `diff` output is already that way:

```
$ diff -u file1.txt file2.txt  
--- file1.txt        2013-07-06 17:44:59.180000000 +0200  
+++ file2.txt        2013-07-06 17:39:53.433000000 +0200  
@@ -1,5 +1,5 @@  
 line1
 line2  
-line3  
+some text
+line4  
 last line
```

I understand the concern, but for blind users I would like to provide special output that doesn't have as much ASCII art in the first place, and we might want to even hide the code suggestions (but keep the message) from that output to avoid overwhelming the user.

**Comment by estebank** [comment]
@petrochenkov I feel the PR is ready for review.

**Comment by petrochenkov** [comment]
The visual part looks fine to me, but I really don't have any stakes in this.

I'll check the technical details in `rustc_erros` some time this week.

File: src/test/ui/suggestions/expected-boxed-future-isnt-pinned.stderr [hunk 4]
```diff
@@ -31,14 +32,15 @@ error[E0308]: mismatched types
...
-   |              help: store this in the heap by calling `Box::new`: `Box::new(x)`
+   |              ^ expected struct `Box`, found type parameter `F`
    |
    = note:      expected struct `Box<dyn Future<Output = i32> + Send>`
            found type parameter `F`
    = note: for more on the distinction between the stack and the heap, read https://doc.rust-lang.org/book/ch15-01-box.html, https://doc.rust-lang.org/rust-by-example/std/box.html, and https://doc.rust-lang.org/std/boxed/index.html
+help: store this in the heap by calling `Box::new`
+   |
+LL |     Pin::new(Box::new(x))
+   |              +++++++++ +
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/suggestions/expected-boxed-future-isnt-pinned.stderr:43]: It adds a single line of output (-3/+4), while gaining us some clarity on what needs to be changed, particularly with the colors added in this PR. This is an uncommon enough error (there are only two tests out of 12K affected by it) that I consider the added clarity to be a clear win, but I am open to having my mind changed.

**Comment by petrochenkov** [comment]
This PR is too large for reviewing it on GitHub.
Could you move all changes in `compiler` outside of `compiler/rustc_errors` to separate PR(s)?
(`rustc_errors` changes can be moved too if they are refactorings not affecting the output, or are orthogonal to other changes.)

File: src/test/ui/suggestions/expected-boxed-future-isnt-pinned.stderr (hunk 4 above)
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/test/ui/suggestions/expected-boxed-future-isnt-pinned.stderr:43]: I mean this whole commit which changes 7 commits. In this particular case it is indeed just a single extra line of output, but this is not the case for all errors. For example src/test/ui/issues/issue-42954.stderr

**Comment by estebank** [comment]
@petrochenkov removed relevant commits only changing output. There are still some commits there that belong more to the exploration of the problem space, but didn't want to remove them until the final output/approach was evaluated in case I needed to go back to earlier code.

File: src/test/ui/suggestions/expected-boxed-future-isnt-pinned.stderr (hunk 4 above)
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/suggestions/expected-boxed-future-isnt-pinned.stderr]: Moving these changes to a new PR, as requested by the reviewer.

File: compiler/rustc_errors/src/lib.rs [hunk 5]
```diff
@@ -187,29 +187,40 @@ impl SubstitutionPart {
...
         use rustc_span::{CharPos, Pos};
 
         fn push_trailing(
             buf: &mut String,
             line_opt: Option<&Cow<'_, str>>,
             lo: &Loc,
             hi_opt: Option<&Loc>,
-        ) {
+        ) -> usize {
+            let mut count = 0;
```
Comment (petrochenkov) [inline compiler/rustc_errors/src/lib.rs:202]:
```suggestion
            let mut line_count = 0;
```
Looks like we are counting lines here and below, but it's far from being clear.

File: compiler/rustc_errors/src/lib.rs [hunk 6]
```diff
@@ -187,29 +187,40 @@ impl SubstitutionPart {
 impl CodeSuggestion {
     /// Returns the assembled code suggestions, whether they should be shown with an underline
     /// and whether the substitution only differs in capitalization.
-    pub fn splice_lines(&self, sm: &SourceMap) -> Vec<(String, Vec<SubstitutionPart>, bool)> {
+    pub fn splice_lines(
+        &self,
+        sm: &SourceMap,
+    ) -> Vec<(String, Vec<SubstitutionPart>, Vec<Vec<(usize, usize)>>, bool)> {
```
Comment (petrochenkov) [inline compiler/rustc_errors/src/lib.rs:193]: It's not clear what `Vec<Vec<(usize, usize)>>` means, probably better to make a struct with named fields for `(usize, usize)`.

File: src/test/ui/borrowck/borrowck-vec-pattern-nesting.stderr [hunk 7]
```diff
@@ -36,12 +36,12 @@ LL |         &mut [_a,
...
-LL |         [_a,
-LL |
-LL |
-LL |
-LL |             ..
-LL |         ] => {
+LL ~         [_a,
+LL +
+LL +
+LL +
```
Comment (petrochenkov) [inline src/test/ui/borrowck/borrowck-vec-pattern-nesting.stderr:42]: Not sure why these empty lines are considered additions.

Comment (estebank) [reply src/test/ui/borrowck/borrowck-vec-pattern-nesting.stderr:42]: That's what the follow up commits that I removed were addresssing: the highlighting mechanism is dump: it doesn't actually attempt to calculate a diff. What it _does_ is mark any whole line that is part of the replacement snippet as an addition, meaning that if the suggestion is in the form of `format!("some addition {}", snippet)`, everything that was part of `snippet` is _also_ marked as an addition. I have some thoughts about how to improve that, but for now I think we can address this best by gradually using more accurate spans.

File: compiler/rustc_errors/src/lib.rs (hunk 6 above)
Comment (estebank) [reply compiler/rustc_errors/src/lib.rs:193]: `Vec< Vec<(usize, usize) >>`:

 - First `Vec` corresponds to each line of the resulting suggestion
 - Second `Vec` corresponds to each highlighted replacement of a multipart suggestion
 - the first `usize` is the start of the pseudo-span to be highlighted
 - the second `usize` is the end of the pseudo-span

Should I also somehow make the `Vec<Vec<_>>` hierarchy clearer? I hoped that the _uses_ would be enough, but maybe embedded comments, at least, would help too?

File: compiler/rustc_errors/src/lib.rs (hunk 5 above)
Comment (estebank) [reply compiler/rustc_errors/src/lib.rs:202]: We are counting the lines that the inserted snippet is adding to the output. Adding a comment and renaming.

File: compiler/rustc_errors/src/emitter.rs [hunk 8]
```diff
@@ -1631,6 +1670,7 @@ impl EmitterWriter {
             // Only show an underline in the suggestions if the suggestion is not the
             // entirety of the code being shown and the displayed code is not multiline.
             if show_underline {
+                let padding: usize = max_line_num_len + 3;
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_errors/src/emitter.rs:1698]: It currently is outside of a `for` loop. I'm pushing it closer to the uses, with the expectation that the naïve perf issue will be optimized away.

File: compiler/rustc_errors/src/lib.rs (hunk 6 above)
Comment (petrochenkov) [reply compiler/rustc_errors/src/lib.rs:193]: A single comment on `fn splice_lines` explaining what it returns should be enough.

**Comment by camsteffen** [comment]
The two-line diff view is added here, but not always used, right? When is it used?

The `^` has an advantage over `+` and `~` in that it has high vertical alignment. So there's a tradeoff. I'm more skeptical of `~`. It has the largest gap from the text above it. And "like a spellchecker" is a little more of a stretch than "plus means add".

I'm wondering if we can have an easy way to "demo" and compare some different options and get more eyes on this, if that is wanted. Viewing the stderr files doesn't really tell the whole story. You could put colored output into a text file so anyone can view it with `less -R demo.txt`.

Sorry if I am over-bikeshedding. 😬

**Comment by estebank** [comment]
> The two-line diff view is added here, but not always used, right? When is it used?

It's used _only_ when there are deletions and the resulting code is a single line.

> I'm more skeptical of ~. It has the largest gap from the text above it.

I chose it because it to avoid reusing `^` and `-` to make it visually distinct while avoiding overloading of meaning, and because of this point that was raised earlier:

> Maybe ~ as an underline for changes? (It's similar to the proofreader's mark for transposition, which is a bit of a stretch but close enough, I think.) [edit: also, I think the existing ^ underline for insertions/changes is probably good enough, and it doesn't have the association with diff that + would]

> I'm wondering if we can have an easy way to "demo" and compare some different options and get more eyes on this

We currently don't have "color correctness" tests in our suite (but we should), it just feels slightly out of scope for this PR.

I also want to eventually make this configurable (maybe with a `-Z` flag at first), but until the code is reviewed and in a relatively good state for merging, adding branching to allow for that just complicates both the code and the reviewer's work.

File: compiler/rustc_errors/src/lib.rs [hunk 9]
```diff
@@ -161,32 +161,77 @@ pub struct SubstitutionPart {
...
+    start: usize,
+    end: usize,
+}
+
+impl SubstitutionPart {
+    pub fn is_addition(&self, sm: &SourceMap) -> bool {
+        !self.snippet.is_empty()
+            && sm
+                .span_to_snippet(self.span)
+                .map_or(self.span.is_empty(), |snippet| snippet.trim().len() == 0)
```
Comment (petrochenkov) [inline compiler/rustc_errors/src/lib.rs:177]:
```suggestion
                .map_or(self.span.is_empty(), |snippet| snippet.trim().is_empty())
```

File: compiler/rustc_errors/src/lib.rs [hunk 10]
```diff
@@ -161,32 +161,77 @@ pub struct SubstitutionPart {
...
+    }
+
+    pub fn is_deletion(&self) -> bool {
+        self.snippet.trim().len() == 0
```
Comment (petrochenkov) [inline compiler/rustc_errors/src/lib.rs:181]:
```suggestion
        self.snippet.trim().is_empty()
```

**Comment by petrochenkov** [comment]
@estebank 
Could you actually make an MCP for this?

I was going to let the bikeshed continue for a few days before r+'ing, but if we are going to do it anyway then why not making it formal.

**Comment by petrochenkov** [comment]
One more observation is that some parts of this PR are independent from others.

I understand that we may want to treat the "make suggestions more diff-like" change as a whole (despite of it also consisting of two largely independent changes - for single-line, and for multi-line suggestions).

However the change from `^` (single-line case) and `|` (multi-line case) to `~` seems entirely independent and not making suggestions more diff-like, it's just a change from one underline symbol to another.

**Comment by petrochenkov** [comment]
Ok, looks like enough time has passed, even without an MCP.
I'm going to r+ this after https://github.com/rust-lang/rust/pull/86754 is merged.

**Comment by petrochenkov** [comment]
r=me after fixing tidy.

**Comment by estebank** [comment]
@bors r=petrochenkov rollup=never p=1

^ This PR has high chances of incurring merge conflicts.

**Comment by estebank** [comment]
@bors r=petrochenkov rollup=never p=1

*shakes fist in nll's general direction*

**Comment by flip1995** [comment]
@estebank blessing the Clippy tests with `./x.py test src/tools/clippy --bless` worked for me. Using `--stage 2` shouldn't be necessary for this, but that shouldn't be the problem here...

Anyway, here's a [gist of the Clippy patch](https://gist.github.com/flip1995/bc4cac6130e651eb2e69f5b6e86ea78b)

**Comment by estebank** [comment]
@flip1995 thanks for checking! Somehow I changed branches some time back _without realizing_ and _that's_ why `--stage 2 --bless` wasn't working for me 🤦‍♂️

@bors r=petrochenkov rollup=never p=1

**Pull request merged**